### PR TITLE
Use nullifying instead of list-map tango

### PR DESCRIPTION
**Pull request by hauleth**
This provides much simpler API in form of

    # With defaults
    with_child_span name, do: :do_something

    # Without defaults
    with_child_span name, %{file: nil, line: nil, function: nil, module: nil} do
      :do_something
    end

### CODE REVIEW

**Comment by hauleth** [comment]
We probably also should allow using keyword list as a 2nd argument to the `with_child_span` instead of map

File: lib/opencensus/trace.ex [hunk 1]
```diff
@@ -1,25 +1,31 @@
...
+            file: unquote(file),
+            function: unquote(function)
+          },
+          unquote(attributes)
+        )
+
       new_span_ctx =
         :oc_trace.start_span(unquote(label), parent_span_ctx, %{
-          :attributes => unquote(computed_attributes)
+          attributes: :maps.filter(fn _k, v -> v != nil end, attributes)
```
Comment (hauleth) [inline lib/opencensus/trace.ex:28]: This change probably should be moved to the core, so user could remove values that are contained in current context as well.

File: test/opencensus/trace_test.exs [hunk 2]
```diff
@@ -0,0 +1,90 @@
+defmodule Opencensus.TraceTest do
+  use ExUnit.Case
+  use ExUnitProperties
+
+  import Opencensus.Trace
+
+  setup do
+    :application.load(:opencensus)
+    :application.set_env(:opencensus, :send_interval_ms, 1)
+    :application.set_env(:opencensus, :reporter, {PidAttributesReporter, self()})
```
Comment (hauleth) [inline test/opencensus/trace_test.exs:10]: This needs to be changed before `0.8.0` of OpenCensus as this have changed to `:reporters` and list.

**Comment by tsloughter** [comment]
They shouldn't be included by default. For the main method of starting a span it shouldn't be by default going outside the spec in what is included. Secondary functions, arguments or options are fine to change the default.

**Review by deadtrickster — CHANGES_REQUESTED** [review]
I don't think this should be merged. We have opt-in, exactly what we need.

**Comment by hauleth** [comment]
@tsloughter I cannot find anything in the spec about disallowing default attributes. We have much more deviations right now from the standard than few default attributes which IMHO become handy when tracing rather than problematic, remember that from list of supported languages I think that only Ruby has capabilities of providing such defaults as Elixir in similar way, C++ and Erlang need preprocessor for that, Java and C# probably would need stack unwinding for getting "call line", in PHP and Python I do not know if this is possible, and Go AFAIK have no such mechanism for providing them at all.

**Comment by tsloughter** [comment]
@hauleth hm, where do we deviate? I know of one spot which I was actually thinking of opening an issue about today,  the spec says to use `start_span` where we use `with_child_span`.

**Comment by hauleth** [comment]
@tsloughter by [using `oc` namespace](https://github.com/census-instrumentation/opencensus-specs/blob/master/NamespaceAndPackage.md). My main question is: if we want to make these values opt-in then why bother anyway? User can provide them on their own if they need it by using `__ENV__`. So if we do not want to make them on by default then we should remove all logic behind them.

**Comment by tsloughter** [comment]
We us `opencensus` as the app name (package name in that context), which is all that document is declaring required.

And really, I figured sticking to the spec strictly for core stuff exposed for each language would limit debates about what we should/shouldn't have.

I like the idea of it being simple to add those attributes is why I like @deadtrickster 's version, no reason to require everyone to define all those themselves or to have everyone create their own macros that do it. It is useful information. Since technically they could always go directly to `:oc_trace` and consider this to be a helper macro I could also see arguing it is separate and ok to go outside the spec.

Anyway, this ended up being a bigger debate than I'd hoped :). Not to put this on @deadtrickster but I don't care either way at this point, hehe.

**Comment by hauleth** [comment]
For me this:

```elixir
with_child_span "child_span", [
  :module,
  %{"attr" => "value"},
  :line
] do
  :do_something
end
```

Looks much more terrible than:

```elixir
with_child_span "child_span", %{
  attr: "value",
  module: __MODULE__,
  line: __ENV__.line
} do
  :do_something
end
```

So "do or do not, there is no try". If we want to make them opt-in then make them fully user-defined, that will greatly reduce amount of code in our app, and current "compile time expansion" is IMHO terrible idea for such simple thing.

**Comment by deadtrickster** [comment]
I solve not only "how to add line number as attribute", but more like "I have number of magic attributes for this context, let me use this and that".

Btw the code that you like more will just work now :-).

**Comment by hauleth** [comment]
@deadtrickster I am fully aware that this will work right now, my point is that do we need that "macro black-magic" in this case at all. If we do not want to provide "default values" by default then why bother with all that cruft? Whole macro then can be rewritten as:

```elixir
defdelegate with_child_span(name, attributes \\ %{}, func), to: :ocp
```

And call it a day, without any macro-hacks. If we can, we should avoid macros at all cost to not introduce troublesome bugs and stack traces. So in this case if we do not want defaults by default, then we should not bother, and just provide thin wrapper over existing functionalities. No macros.

**Comment by tsloughter** [comment]
I still think it should be a macro and not require a function argument, but agree that macro code can be detrimental. So I'm torn. Had hoped it would just be something we could move on from and hoped wouldn't pop up with any problems.

And I just realized there wasn't a `:all` or similar argument to the macro expansion version, which is what I thought would give it an advantage over manually adding line/module/function.

Starting to think we should have the macro with the defaults plus the defdelegate, `defdelegate with_child_span(name, attributes \\ %{}, func), to: :ocp`, which would not have the default attributes since it isn't a macro.

So it is sort of a, "use the macro if you want the defaults, use the function if you don't."

But again, @deadtrickster put work into a nice macro so I don't want to do that unless he is convinced it is the best way, happy to keep it as is and move on as well if this can be settled some other way :)

**Comment by deadtrickster** [comment]
there is :default not :all though :-)

**Comment by hauleth** [comment]
You cannot have both macro and function, and that split would be IMHO the most confusing option of them all, as it would slightly change behaviour depending on the way of calling.

In general good practice is to avoid macros in Elixir code whenever you can, mostly because of the reasons mentioned above (it also moves some matching problems and warnings from the library code to the application code, which is bad).

So anyway, we need to agree whether we want to have defaults, and if so then I believe we should provide symmetrical macro in the Erlang library, or if we do not want such, then we should make this thin wrapper over Erlang functions and call it a day. Personally I like `nil`-clearing approach, as this provides similarities to the `Logger` functionality, so there is some level of familiarity for the user.

**Comment by tsloughter** [comment]
I don't think see how it would be confusing.. You are either using a macro or a function, it is pretty explicit. But I guess Elixir restricts the naming?

**Comment by hauleth** [comment]
@tsloughter you cannot have macro and function with the same name and arty as this will be confusing for the macro expander what is what as it will end in the AST as:

```elixir
{:with_child_span, ctx, [name, attributes, arg]}
```

Where `arg` will be other function definition or keyword list, so without knowing how is macro defined it will not be able to match one or the other.

So my proposals are still the same:

1. Define macro with default attributes by default, or
2. Define delegation for a thin wrapper

IMHO the option 1. is nicer for the UX as user still will be able to:

- go lower level for creating span without any "magic" metadata
- obliterate the metadata by `nil`ing them out by explicit request

The same should be done in Erlang library where we could create macro `?WITH_CHILD_SPAN` that will provide default attributes like this one.

File: test/opencensus/trace_test.exs (hunk 2 above)
Comment (garthk) [reply test/opencensus/trace_test.exs:10]: Yeah, we missed _that_ bus.

**Comment by zachdaniel** [comment]
Would it be kosher to let them globally disable the default metadata, either in their env config or in their tracer module? This doesn’t seem to me like the kind of thing that people would want to disable in a single spot. More likely it’s either all or nothing, IMO.

**Comment by tsloughter** [comment]
I am going to close this so panda stops reminding me about it. But I don't consider the issue resolved, I just think we can wait until `opentelemetry_elixir` to solve it.

**Pull request closed without merging**